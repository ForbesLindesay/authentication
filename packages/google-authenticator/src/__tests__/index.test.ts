import {
  generateSecret,
  getQRCodeSVG,
  getQRCodeURI,
  verifyToken,
  generateToken,
} from '../';

jest.mock('crypto', () => ({
  ...require.requireActual('crypto'),
  randomBytes(count: number, cb: (err: null, buf: Buffer) => void) {
    const seed = Buffer.from(
      'OxhR9el5IAHiwyAuaoWb9I/8+pVCu9zE0uE8sR779wOjDagq+U3lIZX3n2lmqa97S9qSmUnSHCwIuAc/R5xX0wnjG+UT8vCpApqYserXQGMhRZ73lxLv/7Imw4v3gqsHPwOW9YGOZtNf7RsIz/xct3UPILJyiLyhCTIK2CSFhItpycdya3f9kGZ7gDOTbJOmXKIrarFYqE0e4e3iqvqw34nbxvXEmlPQpSf/wqCKENZspIg1xfXcQQoapKBDmHWLKgDRcDXHWelXu0XeQ3gDMBz434HOsnqbh5e9Y32BUMIkZTtXEKihRNxU+0Wpmo4DcojyyJDJdmxffdKTEAiSiLHeEa5OeRgUfvLvRrlDiEQnhC59dAWYmuKKayuxatikA3OMH8IsW0KqeE7YoFHGJaXEl94LfQ8UlCgX/syTdvYTdDRThkhsMZ3dgnkmhyvoFgw74difZGEZmrMY+Cmye1mEu+FVEwEDkd9IZmV9GPqP2kST5cxC1yUdzYz/pPz2a4oOO+RUls4RjWp7ZzCgfBX7yKcpIlOt+a6x3SbM1oFTyzBV5UhinSfHYQsfQbKytL/djFfwBdvZanvJpsIp+uT1Wnkhj/3hrCZsW86DHC6dql2PbwZrB+1dXz//aUsD0hyBcp4AWypSuepv9tsaw6u+accEnvKm/JQyu4NnOSh3CKB2VEOUOuT2Tntg/rOM5RSFTh7jkJ2zpFdLZYb+n4Jwm3RISKlT8/j0jP0+FkbVSKAJYGxIBDZ0GqZgDPplnOCXcPRpWL/i5rt3gUnspQVLU5ezywTSKgyEvQEha/fVDg0pN2tA4jIw/TJRHFdQFIra8W0KGxQeRW5ZX2Pwb0ledFeH91Uz6z5ZbGyYpMBokdA3jT4BJuQIA15LM1BBtx5/Vc++6P5zjv47Uf8EXVdRmiffdyOFFKez7OAZzD8kxZKVL9iY2vqPEe5xjO01nomONaoK4r1EQ+5GVYH/XGD4pzS7heLD5PfJwbmNYprW9y+nOCKNzYb+ebbr4voEDNiu+TImyHNq//K/W4M5omnNsK66w0AnUA44qYE3wzSLUFHet0oAIyeABn2wmojcN2z5xZAbeeGOz6GOmaopLIxvamlvxC0X/VCOKbcWs0eKQyQXWxCzrLJJwJtPcSDhEPlhzr4STdXR7YVcK/u2lTOvpQ7odSPy2Z+VrnzOV/L0CLVbK+1QNl3CZwwuPjFadbY1kxL6o7IoxR/3yb1BR/NzukR8JzXyTQoxmmYKHEs+IQiQRqUYR+tbM4KTXYPrPlAFOtuz4MPt2Lw9JnnjYMQ7FD00DNAuOC55RrIvxWOSwFPWjEc7IJsl',
      'base64',
    );
    const result = Buffer.allocUnsafe(count);
    seed.copy(result, 0, 0, count);
    return cb(null, result);
  },
}));

const TEST_SECRET = 'HMMFD5PJPEQADYWDEAXGVBM36SH7Z6UV';
test('generateSecret', async () => {
  const secret = await generateSecret();
  expect(secret.length).toBe(32);
  expect(secret).toBe(TEST_SECRET);
});

test('getQRCodeSVG', async () => {
  expect(
    await getQRCodeSVG({
      secret: TEST_SECRET,
      label: 'TestApp:user@example.com',
      issuer: 'TestApp',
    }),
  ).toMatchSnapshot();
});

test('getQRCodeURI', async () => {
  expect(
    await getQRCodeURI({
      secret: TEST_SECRET,
      label: 'TestApp:user@example.com',
      issuer: 'TestApp',
    }),
  ).toMatchSnapshot();
});

test('verifyToken', () => {
  // generate a valid token and test it
  const token = generateToken({
    secret: TEST_SECRET,
  });
  expect(
    verifyToken({
      secret: TEST_SECRET,
      token,
    }),
  ).toBe(true);

  // use an old token and check it fails
  expect(
    verifyToken({
      secret: TEST_SECRET,
      token: '157688',
    }),
  ).toBe(false);
});
