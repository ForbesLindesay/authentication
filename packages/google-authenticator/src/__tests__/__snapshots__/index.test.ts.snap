// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getQRCodeSVG 1`] = `
"<svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 49 49\\" shape-rendering=\\"crispEdges\\"><path fill=\\"#ffffff\\" d=\\"M0 0h49v49H0z\\"/><path stroke=\\"#000000\\" d=\\"M4 4.5h7m1 0h1m1 0h1m3 0h3m3 0h3m1 0h3m2 0h1m2 0h1m1 0h7M4 5.5h1m5 0h1m2 0h3m1 0h2m1 0h1m1 0h1m1 0h4m1 0h2m2 0h1m4 0h1m5 0h1M4 6.5h1m1 0h3m1 0h1m2 0h3m1 0h1m1 0h5m2 0h1m1 0h1m1 0h1m4 0h1m2 0h1m1 0h3m1 0h1M4 7.5h1m1 0h3m1 0h1m1 0h2m1 0h1m3 0h3m1 0h3m1 0h2m2 0h1m1 0h3m2 0h1m1 0h3m1 0h1M4 8.5h1m1 0h3m1 0h1m1 0h1m1 0h1m12 0h1m1 0h1m1 0h4m1 0h1m1 0h1m1 0h3m1 0h1M4 9.5h1m5 0h1m1 0h1m1 0h3m3 0h1m2 0h1m2 0h1m2 0h2m1 0h2m1 0h1m2 0h1m5 0h1M4 10.5h7m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h7M12 11.5h1m3 0h1m1 0h6m1 0h4m4 0h2m1 0h1M4 12.5h1m3 0h1m1 0h3m2 0h2m1 0h2m1 0h1m1 0h1m2 0h1m3 0h12m2 0h1M5 13.5h1m1 0h3m1 0h2m2 0h2m1 0h4m5 0h1m1 0h6m1 0h1m1 0h2m3 0h2M5 14.5h8m1 0h1m1 0h2m3 0h2m2 0h1m1 0h3m1 0h2m1 0h1m2 0h8M5 15.5h2m6 0h6m3 0h2m1 0h2m3 0h2m1 0h1m1 0h1m1 0h5m1 0h1M4 16.5h1m1 0h2m1 0h4m1 0h2m4 0h3m1 0h2m7 0h1m2 0h1m1 0h1m5 0h1M7 17.5h3m2 0h1m2 0h2m2 0h2m1 0h2m2 0h1m1 0h1m1 0h1m1 0h1m1 0h9M4 18.5h1m2 0h1m1 0h5m1 0h2m2 0h2m2 0h1m2 0h3m3 0h3m3 0h1m5 0h1M4 19.5h1m3 0h2m5 0h2m1 0h2m3 0h3m1 0h1m1 0h2m3 0h1m3 0h2m1 0h1m1 0h2M4 20.5h1m2 0h1m2 0h1m1 0h5m1 0h5m2 0h1m2 0h4m1 0h1m1 0h1m1 0h1m1 0h1m1 0h1m1 0h2M4 21.5h1m1 0h1m2 0h1m1 0h2m1 0h1m1 0h1m3 0h2m2 0h1m1 0h1m1 0h1m1 0h2m4 0h5m1 0h1M5 22.5h1m2 0h4m2 0h7m1 0h6m1 0h2m1 0h2m1 0h1m5 0h2m1 0h1M7 23.5h2m2 0h1m2 0h3m1 0h1m6 0h1m3 0h1m2 0h3m3 0h2m3 0h1M4 24.5h1m1 0h6m5 0h1m1 0h1m7 0h3m2 0h1m1 0h4M4 25.5h1m1 0h1m1 0h2m1 0h2m2 0h3m4 0h3m1 0h2m2 0h2m1 0h1m3 0h2m1 0h3m1 0h1M4 26.5h5m1 0h2m1 0h2m1 0h1m1 0h2m4 0h4m1 0h3m2 0h1m4 0h2m1 0h1m1 0h1M4 27.5h1m1 0h1m5 0h1m1 0h1m1 0h8m1 0h1m4 0h2m1 0h2m1 0h1m1 0h3m1 0h1M4 28.5h4m1 0h3m1 0h5m3 0h1m1 0h1m1 0h2m2 0h1m1 0h3m3 0h1m1 0h2m2 0h2M6 29.5h2m1 0h1m1 0h1m3 0h2m1 0h1m1 0h3m1 0h4m1 0h1m1 0h2m3 0h1m1 0h1m3 0h1m1 0h1M5 30.5h2m2 0h2m2 0h5m1 0h1m1 0h1m3 0h1m1 0h2m1 0h2m5 0h1m1 0h1m3 0h1M4 31.5h1m1 0h1m1 0h1m3 0h1m5 0h5m2 0h1m1 0h2m1 0h1m2 0h2m2 0h5m1 0h1M8 32.5h1m1 0h2m1 0h1m3 0h2m1 0h4m1 0h1m2 0h1m2 0h3m2 0h1m6 0h2M4 33.5h2m5 0h1m1 0h4m1 0h4m1 0h2m5 0h1m2 0h8m1 0h3M8 34.5h1m1 0h1m1 0h1m1 0h1m3 0h2m1 0h1m2 0h2m3 0h1m1 0h1m1 0h1m4 0h2m1 0h1m1 0h2M8 35.5h1m2 0h3m2 0h3m2 0h2m3 0h1m1 0h1m1 0h3m1 0h1m2 0h5m1 0h1M4 36.5h3m3 0h3m1 0h2m3 0h2m1 0h2m1 0h1m1 0h2m3 0h2m2 0h5m3 0h1M12 37.5h5m7 0h7m2 0h1m1 0h2m3 0h2m1 0h2M4 38.5h7m1 0h1m2 0h1m1 0h5m1 0h2m2 0h2m3 0h1m1 0h1m1 0h1m1 0h1m1 0h2m2 0h1M4 39.5h1m5 0h1m2 0h2m1 0h1m2 0h1m1 0h2m3 0h1m3 0h3m3 0h1m3 0h1m3 0h1M4 40.5h1m1 0h3m1 0h1m1 0h1m2 0h1m2 0h1m1 0h1m1 0h1m1 0h1m1 0h2m1 0h2m5 0h5m2 0h1M4 41.5h1m1 0h3m1 0h1m2 0h1m2 0h2m1 0h1m3 0h2m2 0h1m1 0h1m1 0h2m1 0h2m1 0h1m3 0h1m2 0h1M4 42.5h1m1 0h3m1 0h1m3 0h2m1 0h3m1 0h2m1 0h5m1 0h2m1 0h2m1 0h2m3 0h1m1 0h2M4 43.5h1m5 0h1m4 0h3m1 0h1m2 0h1m2 0h2m6 0h1m2 0h4m2 0h3M4 44.5h7m1 0h7m1 0h2m1 0h1m1 0h2m1 0h1m2 0h1m1 0h1m2 0h1m3 0h1m2 0h1\\"/></svg>
"
`;

exports[`getQRCodeURI 1`] = `"data:image/svg+xml,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%2049%2049%22%20shape-rendering%3D%22crispEdges%22%3E%3Cpath%20fill%3D%22%23ffffff%22%20d%3D%22M0%200h49v49H0z%22%2F%3E%3Cpath%20stroke%3D%22%23000000%22%20d%3D%22M4%204.5h7m1%200h1m1%200h1m3%200h3m3%200h3m1%200h3m2%200h1m2%200h1m1%200h7M4%205.5h1m5%200h1m2%200h3m1%200h2m1%200h1m1%200h1m1%200h4m1%200h2m2%200h1m4%200h1m5%200h1M4%206.5h1m1%200h3m1%200h1m2%200h3m1%200h1m1%200h5m2%200h1m1%200h1m1%200h1m4%200h1m2%200h1m1%200h3m1%200h1M4%207.5h1m1%200h3m1%200h1m1%200h2m1%200h1m3%200h3m1%200h3m1%200h2m2%200h1m1%200h3m2%200h1m1%200h3m1%200h1M4%208.5h1m1%200h3m1%200h1m1%200h1m1%200h1m12%200h1m1%200h1m1%200h4m1%200h1m1%200h1m1%200h3m1%200h1M4%209.5h1m5%200h1m1%200h1m1%200h3m3%200h1m2%200h1m2%200h1m2%200h2m1%200h2m1%200h1m2%200h1m5%200h1M4%2010.5h7m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h7M12%2011.5h1m3%200h1m1%200h6m1%200h4m4%200h2m1%200h1M4%2012.5h1m3%200h1m1%200h3m2%200h2m1%200h2m1%200h1m1%200h1m2%200h1m3%200h12m2%200h1M5%2013.5h1m1%200h3m1%200h2m2%200h2m1%200h4m5%200h1m1%200h6m1%200h1m1%200h2m3%200h2M5%2014.5h8m1%200h1m1%200h2m3%200h2m2%200h1m1%200h3m1%200h2m1%200h1m2%200h8M5%2015.5h2m6%200h6m3%200h2m1%200h2m3%200h2m1%200h1m1%200h1m1%200h5m1%200h1M4%2016.5h1m1%200h2m1%200h4m1%200h2m4%200h3m1%200h2m7%200h1m2%200h1m1%200h1m5%200h1M7%2017.5h3m2%200h1m2%200h2m2%200h2m1%200h2m2%200h1m1%200h1m1%200h1m1%200h1m1%200h9M4%2018.5h1m2%200h1m1%200h5m1%200h2m2%200h2m2%200h1m2%200h3m3%200h3m3%200h1m5%200h1M4%2019.5h1m3%200h2m5%200h2m1%200h2m3%200h3m1%200h1m1%200h2m3%200h1m3%200h2m1%200h1m1%200h2M4%2020.5h1m2%200h1m2%200h1m1%200h5m1%200h5m2%200h1m2%200h4m1%200h1m1%200h1m1%200h1m1%200h1m1%200h1m1%200h2M4%2021.5h1m1%200h1m2%200h1m1%200h2m1%200h1m1%200h1m3%200h2m2%200h1m1%200h1m1%200h1m1%200h2m4%200h5m1%200h1M5%2022.5h1m2%200h4m2%200h7m1%200h6m1%200h2m1%200h2m1%200h1m5%200h2m1%200h1M7%2023.5h2m2%200h1m2%200h3m1%200h1m6%200h1m3%200h1m2%200h3m3%200h2m3%200h1M4%2024.5h1m1%200h6m5%200h1m1%200h1m7%200h3m2%200h1m1%200h4M4%2025.5h1m1%200h1m1%200h2m1%200h2m2%200h3m4%200h3m1%200h2m2%200h2m1%200h1m3%200h2m1%200h3m1%200h1M4%2026.5h5m1%200h2m1%200h2m1%200h1m1%200h2m4%200h4m1%200h3m2%200h1m4%200h2m1%200h1m1%200h1M4%2027.5h1m1%200h1m5%200h1m1%200h1m1%200h8m1%200h1m4%200h2m1%200h2m1%200h1m1%200h3m1%200h1M4%2028.5h4m1%200h3m1%200h5m3%200h1m1%200h1m1%200h2m2%200h1m1%200h3m3%200h1m1%200h2m2%200h2M6%2029.5h2m1%200h1m1%200h1m3%200h2m1%200h1m1%200h3m1%200h4m1%200h1m1%200h2m3%200h1m1%200h1m3%200h1m1%200h1M5%2030.5h2m2%200h2m2%200h5m1%200h1m1%200h1m3%200h1m1%200h2m1%200h2m5%200h1m1%200h1m3%200h1M4%2031.5h1m1%200h1m1%200h1m3%200h1m5%200h5m2%200h1m1%200h2m1%200h1m2%200h2m2%200h5m1%200h1M8%2032.5h1m1%200h2m1%200h1m3%200h2m1%200h4m1%200h1m2%200h1m2%200h3m2%200h1m6%200h2M4%2033.5h2m5%200h1m1%200h4m1%200h4m1%200h2m5%200h1m2%200h8m1%200h3M8%2034.5h1m1%200h1m1%200h1m1%200h1m3%200h2m1%200h1m2%200h2m3%200h1m1%200h1m1%200h1m4%200h2m1%200h1m1%200h2M8%2035.5h1m2%200h3m2%200h3m2%200h2m3%200h1m1%200h1m1%200h3m1%200h1m2%200h5m1%200h1M4%2036.5h3m3%200h3m1%200h2m3%200h2m1%200h2m1%200h1m1%200h2m3%200h2m2%200h5m3%200h1M12%2037.5h5m7%200h7m2%200h1m1%200h2m3%200h2m1%200h2M4%2038.5h7m1%200h1m2%200h1m1%200h5m1%200h2m2%200h2m3%200h1m1%200h1m1%200h1m1%200h1m1%200h2m2%200h1M4%2039.5h1m5%200h1m2%200h2m1%200h1m2%200h1m1%200h2m3%200h1m3%200h3m3%200h1m3%200h1m3%200h1M4%2040.5h1m1%200h3m1%200h1m1%200h1m2%200h1m2%200h1m1%200h1m1%200h1m1%200h1m1%200h2m1%200h2m5%200h5m2%200h1M4%2041.5h1m1%200h3m1%200h1m2%200h1m2%200h2m1%200h1m3%200h2m2%200h1m1%200h1m1%200h2m1%200h2m1%200h1m3%200h1m2%200h1M4%2042.5h1m1%200h3m1%200h1m3%200h2m1%200h3m1%200h2m1%200h5m1%200h2m1%200h2m1%200h2m3%200h1m1%200h2M4%2043.5h1m5%200h1m4%200h3m1%200h1m2%200h1m2%200h2m6%200h1m2%200h4m2%200h3M4%2044.5h7m1%200h7m1%200h2m1%200h1m1%200h2m1%200h1m2%200h1m1%200h1m2%200h1m3%200h1m2%200h1%22%2F%3E%3C%2Fsvg%3E%0A"`;
